### ISSUE TITLE
vmap function does not compile jitted function

### ISSUE THREAD

**Issue by msaroufim** [pytorch/functorch#710]
Looking at a lot of jax code they seem to freely mix together `jit` and `vmap` for efficient vectorization.

So I tried to do the same in functorch but found that using vmap on a function with `@torch.jit.script` is about 20x slower than not using it. I have a simple repro below and that 20x different is removed if I manually call the torch.jit.script function. I guess I would expect `vmap` to do that for me instead of having me warmup all the jitted functions.

```python
import torch
from functorch import vmap
import time 

@torch.jit.script
def double(x):
    return x * 2

def non_jit_double(x):
    return x * 2

x = torch.randn(100000)

# need to warmup the function that will be jitted otherwise 20x performance degredation
y = double(torch.randn(1000))

# this warmum has no impact
y = vmap(double)(x)


tic  = time.time()
y = vmap(double)(x)
toc = time.time()
duration = toc - tic 
print(duration)

# warmup has no impact
y2 = vmap(non_jit_double)(x)

tic = time.time()
y2 = vmap(non_jit_double)(x)
toc = time.time()
duration = toc - tic 
print(duration)
```

EDIT: I called vmap more than once in a for loop and this effect went away

**Comment by msaroufim** [comment]
The interesting part is jit gets optimized after about 10 iterations and not 1 :confounded: 

```python
from functorch import vmap
import time 

@torch.jit.script
def double(x):
    return x * 2

def non_jit_double(x):
    return x * 2

x = torch.randn(100000)

for i in range(20):
    tic = time.time()
    y = vmap(double)(x)
    toc = time.time()
    duration = toc - tic
    print(f"iteration: {i}, duration: {duration}")
```

```
from functorch import vmap
import time 

@torch.jit.script
def double(x):
    return x * 2

def non_jit_double(x):
    return x * 2

x = torch.randn(100000)

for i in range(20):
    tic = time.time()
    y = vmap(double)(x)
    toc = time.time()
    duration = toc - tic
    print(f"iteration: {i}, duration: {duration}")
```

**Comment by Chillee** [comment]
Mmmm, so... this is probably the wrong way to compile through vmap :P It might technically work today, but it's very much not intended usage.

In general we cannot (currently) freely mix vmap and `jit` (or AOTAutograd). It's a WIP but currently doesn't quite work. `aot_function(vmap(f))` *does* (mostly work), but not the other way around.

**Comment by msaroufim** [comment]
> aot_function(vmap(f)) does (mostly work), but not the other way around.

I think this could work, I believe jax recommendation is to jit the vmap and not vmap the jit if I'm not mistaken

Regardless maybe aot_function needs its own decorator `@aot` or `compile` or what have you

**Comment by zou3519** [comment]
> > aot_function(vmap(f)) does (mostly work), but not the other way around.
> 
> I think this could work, I believe jax recommendation is to jit the vmap and not vmap the jit if I'm not mistaken

aot_function(vmap(f)) is jitting the vmap, right?
